File: AboutLayout_ObjC/AboutLayout/MapViewController.m
//
//  MapViewController.m
//  AboutLayout
//
//  Created by NixonShih on 2017/1/19.
//  Copyright © 2017年 Nixon. All rights reserved.
//

#import "MapViewController.h"
#import <MapKit/MapKit.h>

@interface MapViewController () <MKMapViewDelegate>
@property (weak, nonatomic) IBOutlet UILabel *addressLabel;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *loadingView;
@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@end

@implementation MapViewController {
    BOOL isFirst;
    NSTimer *timer;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self prepareMapView];
}

#pragma mark - UI

- (void)prepareMapView {
    _mapView.delegate = self;
}

#pragma mark - geocoder

- (void)showAddressWithCoordinate:(CLLocationCoordinate2D)coordinate {
    
    if (_loadingView.hidden) {
        _loadingView.hidden = false;
        [_loadingView startAnimating];
    }
    
    CLGeocoder *geocoder = [CLGeocoder new];
    CLLocation *location = [[CLLocation alloc] initWithLatitude:coordinate.latitude
                                                      longitude:coordinate.longitude];
    
    [geocoder reverseGeocodeLocation:location completionHandler:^(NSArray<CLPlacemark *> * _Nullable placemarks, NSError * _Nullable error) {
        
        if (_addressLabel.hidden) _addressLabel.hidden = false;
        
        _loadingView.hidden = true;
        
        if (![self checkDataExistWithPlacemarks:placemarks andError:error]) {
            _addressLabel.text = @"未知";
            return;
        }
        
        NSDictionary *dic = placemarks.firstObject.addressDictionary;
        NSString *detail = dic[@"Name"] ? [NSString stringWithFormat:@", %@",dic[@"Name"]] : @"";
        NSString *address = [NSString stringWithFormat:@"%@, %@, %@%@",dic[@"Country"],dic[@"SubAdministrativeArea"],dic[@"City"],detail];
        _addressLabel.text = address;
        
        NSLog(@"%@", dic);
    }];
}

- (BOOL)checkDataExistWithPlacemarks:(NSArray<CLPlacemark *>*)placemarks
                            andError:(NSError*)error {
    
    if (error || placemarks.count == 0) return false;
    
    NSDictionary *dic = placemarks.firstObject.addressDictionary;
    
    if (dic[@"Country"] && dic[@"SubAdministrativeArea"] && dic[@"City"]) return true;
    
    return false;
}

#pragma mark - MKMapViewDelegate

- (void)mapViewDidFinishLoadingMap:(MKMapView *)mapView {
    if (!isFirst) {
        isFirst = true;
        [self showAddressWithCoordinate:mapView.centerCoordinate];
    }
}

- (void)mapView:(MKMapView *)mapView regionDidChangeAnimated:(BOOL)animated {
    
    if (timer) {
        [timer invalidate];
        timer = nil;
    }
    
    timer = [NSTimer scheduledTimerWithTimeInterval:1.f repeats:false block:^(NSTimer * _Nonnull timer) {
        [self showAddressWithCoordinate:mapView.centerCoordinate];
    }];
}

@end

File: AboutLayout_Swift/AboutLayout/MapViewController.swift
//
//  MapViewController.swift
//  AboutLayout
//
//  Created by NixonShih on 2017/1/20.
//  Copyright © 2017年 Nixon. All rights reserved.
//

import UIKit
import MapKit

class MapViewController: UIViewController {
    
    @IBOutlet fileprivate weak var mapView: MKMapView!
    @IBOutlet fileprivate weak var loadingView: UIActivityIndicatorView!
    @IBOutlet fileprivate weak var addressLabel: UILabel!
    fileprivate var isFirst = false
    fileprivate var timer: Timer?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        prepareMapView()
    }
    
    // MARK: - UI
    
    fileprivate func prepareMapView() {
        mapView.delegate = self
    }
    
    // MARK: - geocoder
    
    fileprivate func showAddress(with coordinate: CLLocationCoordinate2D) {
        
        if loadingView.isHidden {
            loadingView.isHidden = false
            loadingView.startAnimating()
        }
        
        let geocoder = CLGeocoder()
        let location = CLLocation(latitude: coordinate.latitude,
                                  longitude: coordinate.longitude)
        
        geocoder.reverseGeocodeLocation(location) { (placemarks, error) in
            
            if self.addressLabel.isHidden { self.addressLabel.isHidden = false }
            
            self.loadingView.isHidden = true
            
            if !self.checkDataExist(with: placemarks, error: error) {
                self.addressLabel.text = "未知"
                return
            }
            
            let dic = (placemarks?.first?.addressDictionary)!
            let detail = dic["Name"] == nil ? "" : ", \(dic["Name"]!)"
            let address = "\(dic["Country"]!), \(dic["SubAdministrativeArea"]!), \(dic["City"]!)\(detail)"
            self.addressLabel.text = address
            
            print("\(dic)")
        }
    }
    
    fileprivate func checkDataExist(with placemarks: [CLPlacemark]?, error: Error?) -> Bool {
        
        if error != nil { return false }
        
        guard let dic = placemarks?.first?.addressDictionary else {
            return false
        }
        
        if dic["Country"] != nil &&
            dic["SubAdministrativeArea"] != nil &&
            dic["City"] != nil {
            return true
        }
        
        return false
    }
}

extension MapViewController: MKMapViewDelegate {
    
    func mapViewDidFinishLoadingMap(_ mapView: MKMapView) {
        
        if !isFirst {
            isFirst = true
            showAddress(with: mapView.centerCoordinate)
        }
    }
    
    func mapView(_ mapView: MKMapView, regionDidChangeAnimated animated: Bool) {
        
        timer?.invalidate()
        
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: false, block: { (aTimer) in
            self.showAddress(with: mapView.centerCoordinate)
        })
    }
}
